harness/determined/pytorch/_pytorch_context.py

```diff
@@ -385,7 +429,7 @@ def backward(
                 be constructed, allowing to compute higher order derivative
                 products. Defaults to ``False``.
         """
-        if self._use_amp:
+        if self._use_apex:
```

Comment:
q: Are we still supporting `apex` even with the native mixed precision? If we got rid of that, we might not even need `step_optimizer` and `backwards` any more.

Reply:
Let's discuss. I don't think we can git rid of apex until we fully support AMP, and I think we should have a transition period to more fully validate the new approach before dropping the old one. But if we can really drop those functions, that would be great. I'll check, but I thought they included a bunch of Horovod logic for DTrain too.

Reply:
So one factor here is that in a straightforward case, users can enable AMP on existing models by simply calling `wrap_caller(automatic=True)`, and wrap their forward pass with `with autocast():`. If we don't have our own backward function anymore, then they also need to scale their own losses, and little by little the "automatic" option is not really any simpler.

That's not a deal breaker, but if we would like to get rid of automatic, IMO we should get rid of the automatic option right now.

Reply:
Just to close the loop on this: seems to me we would still want step_optimizer because it includes distributed training logic. It can be simplified in a post-APEX world when we get there.

Seems to me backward() is not actually required today unless you're using APEX. It's also required if you're using the experimental easy-AMP mode. But other than that, if a user just used PyTorch's backward() natively, I don't think anything would be affected. But we don't say that in the docs: very likely in the future something will come up that'll make us wish we could easily inject stuff in backward(). So I'm going to stick with the status quo here. I think it's best to remove APEX support once we know native AMP is viable and successfully in use with customers. But we should keep backward() as-is.

Reply:
sounds good to me. do we have a plan to know how we decide "native AMP is viable and successfully in use with customers"?

Reply:
We'll just ask :D There's already been some customer input that went into the design choices here.